src/test/java/bio/terra/workspace/service/resource/controlled/ControlledResourceServiceTest.java

```diff
@@ -351,6 +352,60 @@ public void createBqDatasetUndo() throws Exception {
                 userAccessUtils.defaultUserAuthRequest()));
   }
 
+  @Test
+  @DisabledIfEnvironmentVariable(named = "TEST_ENV", matches = bufferServiceDisabledEnvsRegEx)
+  public void deleteBqDataset() throws Exception {
+    Workspace workspace = reusableWorkspace();
+    String projectId = workspace.getGcpCloudContext().get().getGcpProjectId();
+
+    String datasetId = "my_test_dataset";
+    String location = "us-central1";
+
+    ApiGcpBigQueryDatasetCreationParameters creationParameters =
+        new ApiGcpBigQueryDatasetCreationParameters().datasetId(datasetId).location(location);
+    ControlledBigQueryDatasetResource resource =
+        ControlledResourceFixtures.makeDefaultControlledBigQueryDatasetResource()
+            .workspaceId(workspace.getWorkspaceId())
+            .accessScope(AccessScopeType.ACCESS_SCOPE_SHARED)
+            .managedBy(ManagedByType.MANAGED_BY_USER)
+            .datasetName(datasetId)
+            .build();
+
+    ControlledBigQueryDatasetResource createdDataset =
+        controlledResourceService.createBqDataset(
+            resource,
+            creationParameters,
+            Collections.emptyList(),
+            userAccessUtils.defaultUserAuthRequest());
+    assertEquals(resource, createdDataset);
+
+    // Test idempotency of delete by retrying BQ specific steps once.
+    Map<String, StepStatus> retrySteps = new HashMap<>();
+    retrySteps.put(DeleteBigQueryDatasetStep.class.getName(), StepStatus.STEP_RESULT_FAILURE_RETRY);
+    jobService.setFlightDebugInfoForTest(
+        FlightDebugInfo.newBuilder().doStepFailures(retrySteps).build());
```

Comment:
I think it's always good to have the ERROR undo case tested as well with `lastStepFailure`. Otherwise our tests never run the undo path.

Reply:
I'd generally agree, but in this case there is no undo case, only dismal failure. I don't think it's worth testing that

Reply:
I'll push back a little on that. Whether the undos should succeed and the flight end in error or whether it should be a dismal failure is behavior we want to rely on and so should have a test for. That the involved client code (undo return stepresult successful) is simple isn't the same as unimportant.
In this, is it a dismal failure or just an ERROR failure?

Reply:
All 3 steps now return a failure from undo, so this will be a dismal failure. I think that's what we want here because none of these steps can be undone, so we can't guarantee atomicity.

Fair point that we're relying on the lack of undo though, so maybe we should be enforcing this in a test, even if it won't be a very interesting one.